gapic/src/main/com/google/gapid/server/GapisConnection.java

```diff
@@ -117,9 +141,44 @@ public boolean isConnected() {
     }
 
     @Override
+    public synchronized void setLogMonitor(LogMonitor monitor) {
+      if (monitor == logMonitorThread) {
+        return; // No change.
+      }
+      if (logMonitorThread != null) {
+        logMonitorThread.interrupt();
+        logMonitorThread = null;
+      }
+      if (monitor == null) {
+        return;
+      }
+      logMonitorThread = new LogMonitorThread(channel, monitor);
+      logMonitorThread.start();
+    }
+
+    @Override
     public void close() {
       baseChannel.shutdown();
       super.close();
     }
+
+    private static class LogMonitorThread extends Thread {
+      private final Channel channel;
+      private final LogMonitor monitor;
+
+      LogMonitorThread(Channel channel, LogMonitor monitor) {
+        this.channel = channel;
+        this.monitor = monitor;
+      }
+
+      @Override
+      public void run() {
+        Service.GetLogStreamRequest request = Service.GetLogStreamRequest.newBuilder().build();
+        Iterator<Log.Message> it = GapidGrpc.newBlockingStub(channel).getLogStream(request);
+        while (it.hasNext()) {
+          monitor.onLogMessage(it.next());
```

Comment:
This now causes an exception on clean shutdown. Before closing the connection to gapis, the logger thread should be stopped.

I think you'd be better off using a StreamObserver and the async case rather than the blocking case with an extra thread.